### PR TITLE
Add protobq.MessageSaver

### PR DESCRIPTION
**Pull request by odsod**
- feat: add protobq.MessageSaver
- feat: update API to match other proto encoding APIs

### CODE REVIEW

**Review by Jassob — APPROVED** [review]
Looks good to me, handy with package defaults :+1:

File: encoding/protobq/encode.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+	for i := 0; i < m.Fields().Len(); i++ {
+		field := m.Fields().Get(i)
+		if field.IsMap() {
+			continue // TODO: support maps
+		}
+		if oneOf := field.ContainingOneof(); oneOf != nil && !hasAddedOneOf[oneOf.FullName()] {
+			schema = append(schema, &bigquery.FieldSchema{
+				Name: string(oneOf.Name()),
+				Type: bigquery.StringFieldType,
+			})
```
[the remark this answers is not in the document]
Comment (odsod) [reply encoding/protobq/encode.go:54]: The intention is to have a special field that works like an enum, representing which oneof is set (where the enum values are the field names), but I'm thinking this is too non-standard, and I'll remove this in an upcoming PR before the next release.

File: encoding/protobq/encode.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+			protoreflect.Sfixed64Kind,
+			protoreflect.Sint32Kind,
+			protoreflect.Sint64Kind:
+			fieldSchema.Type = bigquery.IntegerFieldType
+		case protoreflect.BoolKind:
+			fieldSchema.Type = bigquery.BooleanFieldType
+		case protoreflect.StringKind:
+			fieldSchema.Type = bigquery.StringFieldType
+		case protoreflect.GroupKind:
+			continue // ignore legacy proto2 group fields
```
[the remark this answers is not in the document]
Comment (odsod) [reply encoding/protobq/encode.go:81]: Most protobuf generators ignore this type of field - I haven't put time into researching exactly what it's used for, but it seems like something obscure that we should not need to document that we don't support.

File: encoding/protobq/encode.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jassob) [reply encoding/protobq/encode.go:54]: And instead only keep the different oneof fields without the mux-field?

[the remark this answers is not in the document]
Comment (odsod) [reply encoding/protobq/encode.go:54]: Exactly - I'll see if I can find a good example message with a oneof field to test on

[the remark this answers is not in the document]
Comment (odsod) [reply encoding/protobq/encode.go]: Found one!

[the remark this answers is not in the document]
Comment (Jassob) [reply encoding/protobq/encode.go]: Are missing fields implicitly set to null in BQ?

[the remark this answers is not in the document]
Comment (odsod) [reply encoding/protobq/encode.go]: Yes, and every field in the infered schema is nullable

**Comment by Jassob** [comment]
I suppose you'll squash the commits before merging?

**Comment by odsod** [comment]
@Jassob Not necessary IMO, they all make sense individually

**Comment by Jassob** [comment]
> @Jassob Not necessary IMO, they all make sense individually

Maybe, but they don't contribute much either. I don't think it is interesting to see eventual missteps in a project, unless they are already committed. As such I prefer to squash commits to reduce the number of commits in a PR and the history, thereby increasing signal-to-noise ratio.

**Comment by odsod** [comment]
I see your point, but in this case I don't think a single squashed commit will make sense given the series of commits, so I will keep them as is.

**Comment by Jassob** [comment]
I don't advocate a single commit, rather that some are squashed and some are left as is, but don't let me stop you.

**Comment by odsod** [comment]
Thanks! :raised_hands:

**Pull request merged**